pytorch_lightning/trainer/trainer.py

```diff
@@ -947,7 +946,7 @@ def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT,
 
         # teardown if necessary (similar calls for spawn plugins are excluded as they have
         # been included at the end of `new_process` functions)
-        if self._distrib_type not in DistributedType.interactive_compatible_types():
+        if not isinstance(self.training_type_plugin, DDPSpawnPlugin):
```

Comment:
- should we add a test to confirm DDP spawn and other parallel plugins all work as expected to avoid potential regressions here in the future?
- for custom plugins, this doesn't generalize unless they all extend DDP Spawn. should the plugin denote that it uses multiprocessing spawn method to create new processes, and that we need this in order to determine the appropriate teardown?

Reply:
yes it does not generalize to plugins that do spawn but are not subclassing our ddp spawn plugin, but this was a tradeoff introduced in #8685. However, the change in this PR covers the case of plugins that don't spawn and therefore get a teardown at the right place.

Reply:
Definitely, agreed that we should fix this first. I meant the custom plug-in support as a follow up beyond the scope of this PR